backend/server.js

```diff
@@ -1,26 +1,136 @@
 import cors from 'cors'
 import mongoose from 'mongoose'
+import crypto from 'crypto'
+import bcrypt from 'bcrypt'
+import listEndpoints from 'express-list-endpoints'
+import dotenv from 'dotenv'
+
+dotenv.config()
 
 const mongoUrl = process.env.MONGO_URL || "mongodb://localhost/authAPI"
-mongoose.connect(mongoUrl, { useNewUrlParser: true, useUnifiedTopology: true })
+mongoose.connect(mongoUrl, { useNewUrlParser: true, useUnifiedTopology: true, useCreateIndex: true })
 mongoose.Promise = Promise
 
-// Defines the port the app will run on. Defaults to 8080, but can be 
-// overridden when starting the server. For example:
-//
-//   PORT=9000 npm start
+const Thought = mongoose.model('Thought', {
+  message: String,
+  createdAt: {
+    type: Date,
+    default: Date.now
+  }
+})
+
+// Here is Schema directly inside the Model. 
+const User = mongoose.model('User', {
+  email: {
+    type: String,
+    required: true,
+    unique: true,
+    trim: true,
+    validate: {
+      validator: (value) => {
+        return /^\w+([\.+-]?\w+)*@\w+([\.-]?\w+)*(\.\w{2,3})+$/.test(value)
+      },
+      message: "Please enter a valid email address"
+    }
+  },
+  password: {
+    type: String,
+    required: true,
+  },
+  accessToken: {
+    type: String,
+    default: () => crypto.randomBytes(128).toString('hex') // when new user is created this will be called and create an accessToken 
+  }
+})
+
+// Middleware function
+const authenticateUser = async (req, res, next) => {
+  const accessToken = req.header('Authorization') 
+
+  try {
+    const user = await User.findOne({ accessToken })
+    if (user) {
+      next()
+    } else {
+      res.status(401).json({ success: false, loggedOut: true, message: 'Not authenticated' })
+    }
+  } catch (error) {
+    res.status(400).json({ success: false, message: 'Invalid request', error })
+  }
+}
+
 const port = process.env.PORT || 8080
 const app = express()
 
-// Add middlewares to enable cors and json body parsing
 app.use(cors())
-app.use(bodyParser.json())
+app.use(express.json())
 
-// Start defining your routes here
 app.get('/', (req, res) => {
-  res.send('Hello world')
+  res.send(listEndpoints(app))
+})
+
+app.get('/thoughts', authenticateUser)
+app.get('/thoughts/', async (req, res) => {
+  try {
+    const thoughts = await Thought.find().sort({ createdAt: "descending" }).limit(20)
+    res.json({ success: true, thoughts})
+  } catch (error) {
+    res.status(404).json({ error: 'Page not found' })
+  }
+})
+
+app.post('/thoughts', authenticateUser)
+app.post('/thoughts/', async (req, res) => {
+  const { message } = req.body // destructure message from request body
+
+  try {
+    await new Thought({ message }).save()
+    const thoughts = await Thought.find().sort({ createdAt: "descending" }).limit(20)
+    res.json({ success: true, thoughts })
+  } catch (error) {
+    res.status(400).json({ message: 'Invalid request', error })
+  }
+})
+
+app.post('/signup', async (req, res) => {
```

Comment:
Just a UX suggestion, signup & signin are such similar terms it could be confusing. Here's a article about it to help explain it better. [https://uxmovement.com/buttons/why-sign-up-and-sign-in-button-labels-confuse-users/](url)